### DISCUSSION TITLE
Backend Clarification on set_tensor and get_tensor

### DISCUSSION THREAD

**Discussion by taronaeo** [ggml-org/llama.cpp#14519] [Q&A] [answered]
The buffer `set_tensor` and `get_tensor` have the following function signatures

```c++
static void ggml_backend_zdnn_buffer_get_tensor(ggml_backend_buffer_t buffer, const ggml_tensor * tensor, void * data, size_t offset, size_t size);

static void ggml_backend_zdnn_buffer_set_tensor(ggml_backend_buffer_t buffer, ggml_tensor * tensor, const void * data, size_t offset, size_t size);
```

Do I have the following understanding correct?
1. `.get_tensor` copies the data from $ggml \ tensor \rightarrow backend \ tensor$
2. `.set_tensor` copies the data from $backend \ tensor \rightarrow ggml \ tensor$
3. `data` parameter is a contiguous buffer allocated by GGML and the actual data for the tensor is found via $data + offset$?
4. when writing the data back from the backend buffer, do we write into `ggml_tensor` or the `data` parameter?

I understand that this was briefly explained [here](https://github.com/ggml-org/llama.cpp/discussions/14394#discussioncomment-13586726) but the wording was not very clear to me.

**Comment by slaren** [comment, marked as answer]
`get_tensor` and `set_tensor` copy data between a user buffer and a backend buffer. If the buffer is stored in the CPU and does not apply any transformations to the tensor data layout, the implementation is simply a `memcpy`:
https://github.com/ggml-org/llama.cpp/blob/c8c4495b8d3a8799e2d46778f993965b0ac1ae43/ggml/src/ggml-backend.cpp#L1908-L1918

If the backend does not store the tensor data in the standard ggml layout, you will need to apply any transformations in `set_tensor` and the inverse in `get_tensor`.

**Reply by taronaeo** [reply]
> `get_tensor` and `set_tensor` copy data between a user buffer and a backend buffer.

Am I right to assume that the user buffer here is the `void * data` parameter in both functions?

I think I'm still confused so let me simplify the English and help me double check if my understanding is correct.

1. `.set_tensor` copies the data from `void * data` parameter into `ggml_tensor->data + offset`
2. `.get_tensor` copies the data from `ggml_tensor->data + offset` into `void * data` parameter

So translating it to my backend,

1. `.set_tensor` copies `void * data` into `zTensor`, and does NOT copy into `ggml_tensor->data + offset` at all
2. `.get_tensor` copies `zTensor` into `void * data`, and does nothing with `ggml_tensor`

Is this understanding correct? Does `ggml_tensor` still expect data to be present when its in a backend?

**Reply by slaren** [reply]
I think there is some confusion regarding what `tensor->data` means. `tensor->data` is a backend-specific pointer to device memory. For host buffers, it is just a pointer to host memory, but for other backends like CUDA, it may be a pointer into device memory that can only be accessed via functions like `cudaMemcpy`:
https://github.com/ggml-org/llama.cpp/blob/c8c4495b8d3a8799e2d46778f993965b0ac1ae43/ggml/src/ggml-cuda/ggml-cuda.cu#L604-L618

For some backends, it may even be interpreted as an offset within a buffer object, and not as an absolute address.

Tensors are allocated in a ggml-backend buffer, and the address range of the buffer goes from the value returned by `ggml_backend_buffer_get_base()`, up to `ggml_backend_buffer_get_base() + ggml_backend_buffer_get_size()`. The value of `tensor->data` is set automatically by the ggml-backend subsystem from an address within that range. For views and temporary tensors in a compute buffer, many tensors may share the same address. Your backend should be able to work with buffers allocated in this way, at least for the generic buffer type. It is possible to have a buffer type exclusively for weights, but that's more complicated.

**Reply by taronaeo** [reply]
I'm still trying to internalise and map this out against the IBM zDNN interface but I have an interim question. I notice that `.set_tensor` is correctly called for every tensor, but `.get_tensor` does not seem to be called at all. I noticed this comment in the opencl codebase as well https://github.com/ggml-org/llama.cpp/blob/67d1ef23c68e1e09dc6d7423d1ef5fc0ea56ed5e/ggml/src/ggml-opencl/ggml-opencl.cpp#L2850-L2855

Does this mean that `.get_tensor` will only be called just before activation functions like softmax? I have the tensor transformation code from zTensor to normal values currently within `.get_tensor` but it needs to be called before saving the values back to the buffer.

**Reply by slaren** [reply]
`get_tensor` is typically only called for obtaining the result of the computation, or to copy a tensor to a different backend.

**Reply by taronaeo** [reply]
Does this mean that immediately after every computation e.g., matmul, `get_tensor` will be called to (in my case), transform the data back and save into the buffer?

I'm not 100% certain yet but I don't think I see `get_tensor` being called after every computation and its throwing weird values because the data is not transformed back... I had to manually add the transform code within the compute routine to have it play nicely

**Reply by slaren** [reply]
No, users call `get_tensor` when they want to get the result of a graph evaluation, typically on the last tensor of the graph. For example, llama.cpp uses `get_tensor` to obtain the logits.

**Reply by taronaeo** [reply]
Oh my god, thats why the values are not being transformed back.

If;
1. `init_tensor` is used on every tensor to create things like `zTensor` and store it into `tensor->extra`,
2. `set_tensor` is used on every tensor to copy data into the backend i.e., into `tensor->extra->ztensor`,

where should we expect to transform values from `tensor->extra->ztensor` on every tensor back to normal values? I don't see any accompanying function that can help this other than to transform the value back within the compute routine, which likely means overhead 😅

**Reply by slaren** [reply]
Please head to https://github.com/ggml-org/ggml and take a look at the examples. Familiarize yourself with the ggml API before trying to implement a backend.
